controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java

```diff
@@ -102,12 +143,31 @@ protected synchronized boolean tryAcquire(int permit) {
         }
 
         @Override
-        protected synchronized boolean tryRelease(int permit) {
-            final Object value = permitHolder.get();
-            if (value == null) {
+        protected int tryAcquireShared(final int permit) {
+            if (permitHolder != null) {
+                return -1;
+            }
+            if (compareAndSetState(0, 1)) {
+                return 1;
+            } else {
```

Comment:
If this thread is racing with one calling tryAcquire, at this point we don't the status of 'permit'. Perhaps the other thread has set state to 1, but hasn't set permit yet. Or perhaps by now it has set permit. We just know it hadn't set state to 1 yet when our thread got to L150.

Deferring the if (permitHolder != null) check until we get to this else block avoids that, but now we're racing with a thread calling tryRelease. Perhaps checking both places will help; I haven't thought that through yet. :)

I think these issues are why the methods were synchronized.